include/hipacc/AST/ASTFuse.h

```diff
@@ -41,14 +41,14 @@
 #include <clang/Sema/Ownership.h>
 #include <llvm/ADT/SmallVector.h>
 
-#include "hipacc/Analysis/KernelStatistics.h"
-#include "hipacc/Analysis/HostDataDeps.h"
-#include "hipacc/AST/ASTNode.h"
-#include "hipacc/AST/ASTTranslate.h"
-#include "hipacc/Config/CompilerOptions.h"
-#include "hipacc/Device/Builtins.h"
-#include "hipacc/DSL/ClassRepresentation.h"
-#include "hipacc/Vectorization/SIMDTypes.h"
+#include "KernelStatistics.h"
+#include "HostDataDeps.h"
+#include "ASTNode.h"
+#include "ASTTranslate.h"
+#include "CompilerOptions.h"
+#include "Builtins.h"
+#include "ClassRepresentation.h"
+#include "SIMDTypes.h"
```

Comment:
Question: Why go for the local-include-approach? To be honest, I'm more in favor of the absolute path approach, going even further than before and including by using the full path from the workspace root e.g., `include/hipacc/AST/ASTNode.h`. This way its more or less guaranteed to have a unique path for each header.

Reply:
correct, but the old approach has one disadvantage: if i mistakenly introduce some circular dependencies on the header from two targets. CMake during generation phase will not be able to detect this. If we keep include local to targets, the dependency is modeled explicitly. Then during Cmake generation this can already be detected.
This is the main reason i switched to local-include-approach. Correct me if i am wrong

Reply:
@oreiche you mean to include `hipacc/AST/ASTNode.h`?
This makes sense.

Reply:
Let me rephrase it if it was not clear. let's say i have target A and B. 
- If i use global include (old approach), i can do:
   in targetA.h, #include"hipacc/targetB/targetB.h" and
   in targetB.h, #include"hipacc/targetA/targetA.h".

   Then, there is no way to build it. But "cmake .." will still run successfully. although build will fail later.

- if i use local-include-approach, 
   in targetA.h, To be able to #include"targetB.h", I need to target_link_library(targetA, private targetB), and
   in targetB.h, To be able to #include"targetA.h", I need to target_link_library(targetB, private targetA),

  Then "cmake .." will NOT run, because of something like "The inter-target dependency graph contains the following strongly connected component (cycle):". Then you will get a quite clear idea of which targets are the problem. 

I agree at this stage Hipacc does not necessary need it. It is not that complicated anyway. But it's about reducing the chance to introduce errors.

Reply:
@qiao-bo Circular includes should be mitigated by include guards anyway. The problem you are describing sounds like its cause by a naming conflict?

@richardmembarth No, actually I really meant `include/hipacc/...`. Its a matter of taste, but I was recently told that the "Google way" is to include project-owned non-public headers by their "absolute path" in the project (meaning relative path from project's root). This way, you have a unique path for every include.

Example structure:
~~~
include/foo/bar.hpp
src/baz.hpp
src/baz.cpp
~~~

In `baz.cpp`:
~~~cpp
#include "foo/bar.hpp" // include public header, but could also be "include/foo/bar.hpp", as it is project owned
#include "src/baz.hpp" // include private header
~~~

And the include guards in the headers should encode the header path to some degree to be unique as well. (e.g., `SRC_BAZ_HPP` and `INCLUDE_FOO_BAR_HPP`)

Reply:
I'm a fan of `foo/bar.hpp` instead of `include/foo/bar.hpp`.
LLVM handles this the same way.

Reply:
Note that `ASTNode.h` is not a public header, so according to my comment above, I would always include via its full path.

Reply:
@richardmembarth `foo/bar.hpp` is also fine for me.

@qiao-bo I really do like the explicitness you try to introduce here. However, by this approach, you kind of mix linking libraries and including headers. In many cases, you will end up with circular header includes, that are prevented by include guards and might need some forward declarations. After all, its perfectly fine to have two libraries depending on each other include-wise, which would not be possible with your approach :/

Reply:
ok, thanks for the discussion. I will change back to the "foo/bar.hpp" style, with include guards as well.